sklearn/tests/test_common_non_default.py

```diff
@@ -0,0 +1,308 @@
+import pprint
+import itertools
+from inspect import signature
+from typing import Optional, List, Dict, Any
+
+from sklearn.base import is_regressor
+from sklearn.datasets import load_iris
+from sklearn.tree._classes import BaseDecisionTree
+from sklearn.ensemble._forest import BaseForest
+from sklearn.ensemble._gb import BaseGradientBoosting
+from sklearn.utils import all_estimators
+from sklearn.utils.estimator_checks import (
+    _enforce_estimator_tags_y,
+    parametrize_with_checks,
+)
+from sklearn.utils._testing import ignore_warnings
+
+
+categorical_params = [
+    "solver",
+    "algorithm",
+    "loss",
+    "strategy",
+    "selection",
+    "criterion",
+    "multi_class",
+    "kernel",
+    "affinity",
+    "linkage",
+    "metric",
+    "init",
+    "eigen_solver",
+    "initial_strategy",
+    "imputation_order",
+    "encode",
+    "learning_method",
+    "method",
+    "fit_algorithm",
+    "norm",
+    "svd_solver",
+    "order",
+    "mode",
+    "output_distribution",
+]
+bool_params = [
+    "fit_prior",
+    "fit_intercept",
+    "positive",
+    "normalize",
+    "dual",
+    "average",
+    "shuffle",
+    "whiten",
+    "path_method",
+    "include_bias",
+    "interaction_only",
+    "standardize",
+    "with_centering",
+    "with_scaling",
+    "with_mean",
+    "with_std",
+]
+
+
+class FakeParam(str):
+    """Fake string parameter
+
+    This object always returns False when compared to other values,
+    but remember values it was compared with. It is used to determine
+    valid values for a categorical parameter.
+
+    Examples
+    --------
+    >>> fp = FakeParam()
+    >>> isinstance(fp, str)
+    True
+    >>> fp
+    fake-param
+    >>> fp == 'a'
+    False
+    >>> fp.values
+    {'a'}
+    """
+
+    def __init__(self):
+        self.values = set()
+
+    def __repr__(self):
+        return "fake-param"
+
+    def __hash__(self):
+        return hash(str(self))
+
+    def __eq__(self, other):
+        self.values.add(other)
+        return False
+
+
+def detect_valid_categorical_params(
```

Comment:
would it make sense to do this by introducing type hints for categorical params? Or by parsing the docstring?

Reply:
For type annotations we would need [PEP 586](https://www.python.org/dev/peps/pep-0586/) for literal types and backport it for Python <3.8. That's probably indeed the best solution long term.

I'm not very keen on parsing docstrings, as that can break easily and is less reliable than the current apporach. For instance, for estimator taking metrics names as input, all valid names are not always listed, sometimes it's just a link to the user guide.

Reply:
FWIW the [`typing-extensions`](https://pypi.org/project/typing-extensions/) package is an official source of backports too.

And Hypothesis' introspection can inspect `typing` or `typing_extensions.Literal[...]` and automatically sample from the parameters :grin:

Reply:
We are working toward having typing in https://github.com/scikit-learn/scikit-learn/pull/17799

Moving forward we are going to run into invalid parameter combinations when sampling from the parameter space. Libraries such as [ConfigSpace](https://automl.github.io/ConfigSpace/master/User-Guide.html#rd-example-forbidden-clauses) uses a combination of conditional clauses and forbidden clauses to help define the relationships between parameters.

Reply:
Thanks for the references! I think using hypothesis to sample this parameter space long term could be interspersing, as it could also generalize to float parameters.